### PR TITLE
fix: improve types for row grouping

### PR DESCRIPTION
**Pull request by spike-rabbit**
**What kind of change does this PR introduce?** (check one with "x")

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")

- [x] Yes, but the note is already merged
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: projects/ngx-datatable/src/lib/types/group.events.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { Group } from "./group.type";
+
+export interface GroupToggleEvent<TRow> {
```
Comment (timowolf) [inline projects/ngx-datatable/src/lib/types/group.events.ts:3]: Rebase and put in public types.

**Review by timowolf — CHANGES_REQUESTED** [review]
Thanks, this makes sense. I added some comments. Maybe we can merge 2 types. Rebase and place to the new types to public or internal and if possible, write some documentation to the event types.

**Comment by timowolf** [comment]
@spike-rabbit Back to you :-)

**Review by timowolf — CHANGES_REQUESTED** [review]
No other review, find my comments. I pressed the re-request review button by accident.

File: projects/ngx-datatable/src/lib/components/body/body-row-wrapper.component.ts [hunk 2]
```diff
@@ -86,8 +86,7 @@ export class DataTableRowWrapperComponent<TRow = any> implements DoCheck, OnInit
 
   @Input() set rowIndex(val: number) {
     this._rowIndex = val;
-    this.rowContext.rowIndex = val;
-    this.groupContext.rowIndex = val;
+    (this.rowContext ?? this.groupContext).rowIndex = val;
```
[the remark this answers is not in the document]
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/components/body/body-row-wrapper.component.ts:89]: I am against merging. Those types just share two properties. The others are different.
```ts
export interface GroupContext<TRow> {
  group: Group<TRow>;
  expanded: boolean;
  rowIndex: number;
}

export interface RowDetailContext<TRow> {
  row: TRow;
  expanded: boolean;
  rowIndex: number;
  disableRow$?: Observable<boolean>;
}
```

They are emitted by different directives (`DatatableGroupHeaderTemplateDirective` and `DatatableRowDetailTemplateDirective`) so merging them would make them bad as we have properties in there which are never used in certain context. 

Sure, the code looks bad this way, but we need to find a better solution in another MR.  Merging the interface is far more worse in my opinion as this has a negative impact on consumer.

File: projects/ngx-datatable/src/lib/types/group.events.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  type: 'group';
+  value: Group<TRow>;
+}
+
+export interface AllGroupsToggleEvent {
```
[the remark this answers is not in the document]
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/types/group.events.ts:8]: It is part of the `GroupToggleEvents` . The `AllGroupsToggleEvent` type is indirectly used in `DatatableGroupHeaderDirective#expandAllGroups`

File: projects/ngx-datatable/src/lib/types/group.events.ts (hunk 1 above)
Comment (spike-rabbit) [reply projects/ngx-datatable/src/lib/types/group.events.ts]: done

**Comment by timowolf** [comment]
@spike-rabbit I think the body component should also make use of the new toggle types.

**Comment by spike-rabbit** [comment]
No longer needed

**Pull request closed without merging**